src/java/com/rapleaf/jack/queries/GenericQuery.java

```diff
@@ -0,0 +1,147 @@
+import com.rapleaf.jack.ModelField;
+import com.rapleaf.jack.ModelWithId;
+
+public class GenericQuery {
+
+  private final BaseDatabaseConnection dbConnection;
+  private Class<? extends ModelWithId> mainModel;
+  private final List<JoinCondition> joinConditions;
+  private final List<WhereConstraint> whereConstraints;
+  private final List<OrderCriterion> orderCriteria;
+  private final Set<ModelField> selectedModelFields;
+  private Optional<LimitCriterion> limitCriteria;
+
+  private GenericQuery(BaseDatabaseConnection dbConnection) {
+    this.dbConnection = dbConnection;
+    this.mainModel = null;
+    this.joinConditions = Lists.newArrayList();
+    this.whereConstraints = Lists.newArrayList();
+    this.orderCriteria = Lists.newArrayList();
+    this.selectedModelFields = Sets.newHashSet();
+    this.limitCriteria = Optional.absent();
+  }
+
+  public static GenericQuery create(BaseDatabaseConnection dbConnection) {
+    return new GenericQuery(dbConnection);
+  }
+
+  public GenericQueryBuilder from(Class<? extends ModelWithId> model) {
+    this.mainModel = model;
+    return new GenericQueryBuilder(dbConnection, this);
+  }
+
+  List<WhereConstraint> getWhereConstraints() {
+    return whereConstraints;
+  }
+
+  Set<ModelField> getSelectedModelFields() {
+    return selectedModelFields;
+  }
+
+  void addJoinCondition(JoinCondition joinCondition) {
+    this.joinConditions.add(joinCondition);
+  }
+
+  void addWhereCondition(WhereConstraint whereConstraint) {
+    this.whereConstraints.add(whereConstraint);
+  }
+
+  void addOrderCondition(OrderCriterion orderCriterion) {
+    this.orderCriteria.add(orderCriterion);
+  }
+
+  void addLimitCondition(LimitCriterion limitCriterion) {
+    this.limitCriteria = Optional.of(limitCriterion);
+  }
+
+  void addSelectedModelField(ModelField modelField) {
+    this.selectedModelFields.add(modelField);
+  }
+
+  boolean isOrderedQuery() {
+    return !orderCriteria.isEmpty();
+  }
+
+  String getSqlStatement() {
+    StringBuilder statement = new StringBuilder();
+
+    statement.append(getSelectClause())
+        .append(getJoinClause())
+        .append(getWhereClause());
+
+    if (isOrderedQuery()) {
+      statement.append(getOrderClause());
+      statement.append(getLimitClause());
+    }
+
+    return statement.toString();
+  }
+
+  private String getSelectClause() {
+    StringBuilder clause = new StringBuilder("SELECT ");
+
+    if (selectedModelFields.isEmpty()) {
+      clause.append("*");
+    } else {
+      Iterator<ModelField> it = selectedModelFields.iterator();
+      while (it.hasNext()) {
+        clause.append(it.next().getSqlKeyword());
+        if (it.hasNext()) {
+          clause.append(", ");
+        }
+      }
+    }
+
+    String tableName = Utility.getTableName(mainModel);
+    return clause.append(" FROM ").append(tableName).append(" ").toString();
+  }
+
+  private String getJoinClause() {
+    return getClause(joinConditions, "", " ");
```

Comment:
This implementation doesn't support self-joins where the same table appears multiple times. In that case you need to be able to rename tables you are selecting from. Similarly we might want to be able to rename columns we select (at least once we add support for group by).